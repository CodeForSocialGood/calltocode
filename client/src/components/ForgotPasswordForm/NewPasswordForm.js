import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Field, reduxForm } from 'redux-form'
import PropTypes from 'prop-types'

import SignupValidator from '../SignupForm/SignupValidator'
import ValidationPopup from '../SignupForm/ValidationPopup'
import styles from './ForgotPasswordForm.scss'

class NewPasswordForm extends Component {
  renderPassword (field) {
    const passClasses = `${styles.inputPassword} 
                         ${field.meta.pristine ? '' : (field.meta.error ? styles.error : styles.valid)}`
    return (
      <div className={ styles.inputPasswordContainer }>
        <input
          className={ passClasses }
          placeholder="New Password"
          type="password"
          {...field.input} />
        <Field
          name="popup"
          component={ValidationPopup}
          active={field.meta.active}
          error={field.meta.error} />
      </div>
    )
  }

  render () {
    const { handleSubmit } = this.props

    return (
      <form className={styles.form} onSubmit={handleSubmit}>
        <h1 className={styles.title}>New Password</h1>
        <h3>Set your new password</h3>
        <Field
          name="password"
          component={this.renderPassword}
          validate={SignupValidator.validatePassword} />
        <button className={styles.buttonSubmit} type="submit">
          Change Password
        </button>
      </form>
    )
  }
}

NewPasswordForm.propTypes = {
  changePassword: PropTypes.func,
  handleSubmit: PropTypes.func
}

const NewPasswordFormRedux = reduxForm({
  form: 'ForgotPasswordForm', // <------ same form name
  destroyOnUnmount: false, // <------ preserve form data
  forceUnregisterOnUnmount: true // <------ unregister fields on unmount
})(NewPasswordForm)

export default connect(null)(NewPasswordFormRedux)
