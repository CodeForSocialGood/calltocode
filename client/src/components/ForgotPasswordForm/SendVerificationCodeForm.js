import React from 'react'
import {Field, reduxForm} from 'redux-form'
import styles from './ForgotPasswordForm.scss'
import PropTypes from 'prop-types'

function EmailField ({ input }) {
  return (
    <input
      className={styles.inputEmail}
      placeholder="Email"
      {...input} />
  )
}

const SendVerificationCodeForm = props => {
  const {handleSubmit} = props
  return (

    <form onSubmit={handleSubmit}
      className={styles.form}>

      <h1
        className={styles.title}>
            Forgot Password?
      </h1>

      <h3>{"Let's get you a new one!"}</h3>

      <Field
        name="email"
        component={EmailField} />

      <button
        className={styles.buttonSubmit}
        type="submit">
            Send Verification Code
      </button>
    </form>
  )
}

EmailField.propTypes = {
  input: PropTypes.object
}
SendVerificationCodeForm.propTypes = {
  handleSubmit: PropTypes.func
}

export default reduxForm({
  form: 'ForgotPasswordForm', // <------ same form name
  destroyOnUnmount: false, // <------ preserve form data
  forceUnregisterOnUnmount: true // <------ unregister fields on unmount
})(SendVerificationCodeForm)
