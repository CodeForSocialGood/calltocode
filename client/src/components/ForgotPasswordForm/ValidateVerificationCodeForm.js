import React from 'react'
import {Field, reduxForm} from 'redux-form'
import styles from './ForgotPasswordForm.scss'
import PropTypes from 'prop-types'

function CodeField ({ input }) {
  return (
    <input
      className={styles.inputEmail}
      placeholder="Code"
      {...input} />
  )
}

const ValidateVerificationCodeForm = props => {
  const {handleSubmit} = props
  return (

    <form onSubmit={handleSubmit}
      className={styles.form}>

      <h1
        className={styles.title}>
      Forgot Password?
      </h1>

      <h3>{"Let's validate your code!"}</h3>

      <Field
        name="code"
        component={CodeField} />

      <button
        className={styles.buttonSubmit}
        type="submit">
      Validate code
      </button>
    </form>
  )
}

CodeField.propTypes = {
  input: PropTypes.object
}
ValidateVerificationCodeForm.propTypes = {
  handleSubmit: PropTypes.func
}

export default reduxForm({
  form: 'ForgotPasswordForm', // <------ same form name
  destroyOnUnmount: false, // <------ preserve form data
  forceUnregisterOnUnmount: true // <------ unregister fields on unmount
})(ValidateVerificationCodeForm)
