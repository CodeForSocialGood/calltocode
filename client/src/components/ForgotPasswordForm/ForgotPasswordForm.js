import React, { Component } from 'react'
import { reduxForm, SubmissionError } from 'redux-form'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { push } from 'react-router-redux'
import AuthActionCreator from '../../actions/auth'
import SendVerificationCodeForm from './SendVerificationCodeForm'
import ValidateVerificationCode from './ValidateVerificationCodeForm'
import NewPasswordForm from './NewPasswordForm'
import apiOptionsFromState from '../../api/lib/apiOptionsFromState'
import forgotPasswordApiClient from '../../api/forgotPassword'
import styles from './ForgotPasswordForm.scss'

class ForgotPasswordForm extends Component {
  constructor (props) {
    super(props)
    this.nextPage = this.nextPage.bind(this)
    this.validate = this.validate.bind(this)
    this.changePass = this.changePass.bind(this)
    this.state = {
      page: 1
    }
  }

  nextPage ({email}) {
    return this.props.sendValidationCode(email).then(() => {
      this.setState({page: this.state.page + 1})
    }).catch((error) => { throw new SubmissionError({ code: error.name, _error: error.message }) })
  }

  async validate (values, dispatch) {
    const _error = 'Incorrect code, please try again!'
    const apiOptions = apiOptionsFromState()
    const response = await forgotPasswordApiClient.validateCode(apiOptions, this.props.email, values.code)
    if (response.status === 200) {
      this.setState({page: this.state.page + 1})
    } else {
      throw new SubmissionError({ code: response.statusText, _error })
    }
  }

  changePass (values, dispatch) {
    return this.props.changePassword(this.props.email, values.password).then(() => {
      dispatch(push('/'))
    }).catch(error => {
      throw new SubmissionError({code: error.name, _error: error.message})
    })
  }

  render () {
    const {page} = this.state
    const {error} = this.props
    console.log('error', error)
    return (
      <div>
        {page === 1 && <SendVerificationCodeForm onSubmit={this.nextPage} />}
        {page === 2 &&
        <div>
          <ValidateVerificationCode onSubmit={this.validate}/>
        </div>}
        {page === 3 &&
        <div>
          <NewPasswordForm email={this.props.email} onSubmit={this.changePass}/>
        </div>}
        <div className={styles.errorContent}>
          {error}
        </div>
      </div>

    )
  }
}

ForgotPasswordForm.propTypes = {
  sendValidationCode: PropTypes.func,
  validateCode: PropTypes.func,
  email: PropTypes.string,
  handleSubmit: PropTypes.func,
  error: PropTypes.any,
  changePassword: PropTypes.func
}

const mapDispatchToProps = {
  sendValidationCode: AuthActionCreator.sendValidationCode,
  changePassword: AuthActionCreator.changePassword
}

function mapStateToProps (state) {
  return {
    email: state.forgotPass.email
  }
}

const ForgotPasswordFormRedux = reduxForm({
  form: 'ForgotPasswordForm',
  onSubmitSuccess: (result, dispatch) => {
    dispatch(push('/'))
  }
})(ForgotPasswordForm)

export default connect(mapStateToProps, mapDispatchToProps)(ForgotPasswordFormRedux)
