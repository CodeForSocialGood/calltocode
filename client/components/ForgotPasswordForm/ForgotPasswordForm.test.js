/* eslint-env jest */
import React from 'react'
import {shallow} from 'enzyme'

import {ForgotPasswordForm} from './ForgotPasswordForm'
let mockProps
let mockState

beforeEach(() => {
  mockProps = {
    sendValidationCode: () => {},
    validateCode: () => {},
    email: null,
    handleSubmit: () => {},
    error: 'this is some error text',
    changePassword: () => {}
  }
  mockState = {
    page: null
  }
})

describe('<ForgotPasswordForm />', () => {
  describe('page === 1', () => {
    beforeEach(() => { mockState.page = 1 })

    test('component renders correctly', () => {
      const component = shallow(<ForgotPasswordForm {...mockProps} />)
      component.setState(mockState)
      expect(component.update()).toMatchSnapshot()
    })
  })

  describe('page === 2', () => {
    beforeEach(() => { mockState.page = 2 })

    test('component renders correctly', () => {
      const component = shallow(<ForgotPasswordForm {...mockProps} />)
      component.setState(mockState)
      expect(component.update()).toMatchSnapshot()
    })
  })

  describe('page === 3', () => {
    beforeEach(() => { mockState.page = 2 })

    test('component renders correctly', () => {
      const component = shallow(<ForgotPasswordForm {...mockProps} />)
      component.setState(mockState)
      expect(component.update()).toMatchSnapshot()
    })
  })
})
