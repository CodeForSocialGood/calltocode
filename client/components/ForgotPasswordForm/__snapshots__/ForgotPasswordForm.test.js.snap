// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPasswordForm /> page === 1 component renders correctly 1`] = `
<div>
  <ReduxForm
    email=""
    onChangeEmail={[Function]}
    onSubmit={[Function]}
  />
  <div
    className="errorContent"
  >
    this is some error text
  </div>
</div>
`;

exports[`<ForgotPasswordForm /> page === 2 component renders correctly 1`] = `
<div>
  <div>
    <ReduxForm
      code=""
      onChangeCode={[Function]}
      onSubmit={[Function]}
    />
  </div>
  <div
    className="errorContent"
  >
    this is some error text
  </div>
</div>
`;

exports[`<ForgotPasswordForm /> page === 3 component renders correctly 1`] = `
<div>
  <div>
    <ReduxForm
      code=""
      onChangeCode={[Function]}
      onSubmit={[Function]}
    />
  </div>
  <div
    className="errorContent"
  >
    this is some error text
  </div>
</div>
`;
